yt/fields/field_type_container.py

```diff
@@ -112,46 +115,54 @@ def __contains__(self, obj):
                 return True
         return False
 
-    def _ipython_display_(self):
-        import ipywidgets
-        from IPython.display import Markdown, display
-
-        names = dir(self)
-        names.sort()
-
-        def change_field(_ftype, _box, _var_window):
-            def _change_field(event):
-                fobj = getattr(_ftype, event["new"])
-                _box.clear_output()
-                with _box:
-                    display(
-                        Markdown(
-                            data="```python\n"
-                            + textwrap.dedent(fobj.get_source())
-                            + "\n```"
+    if IPYWIDGETS_ENABLED:
```

Comment:
I really don't like this conditional in the class body.  Would you be open to changing it into a decorator or something?

Reply:
I don't mind a decorator, but how should it behave ? We originally promoted ipywidgets as a hard dependency so this method wouldn't raise if it was missing, but I don't see what else a decorator could do (except maybe change the function name at runtime so it would effectively behave the same as my conditional... just more obfuscated)

Reply:
I'm really feeling anxious about the conditional inside the class body -- in general we don't have places where we evaluate code (other than decorators and traitlet initializers) inside class bodies.  It's not against the rules to do it, per se, but it feels out of place and the additional indent gets in the way; we often have monkey-patched in the past.

But in looking at it, I'm starting to think that we don't need to conditionally define it at all.  If `_ipython_display_` is being called, it's because it's being evaluated inside an ipython environment (or someone is deliberately trying to do something.)  So at that point, I think we could evaluate whether `IPYWIDGETS_ENABLED` is true (inside the function body) and if not, fall back to `_repr_`.

Reply:
> So at that point, I think we could evaluate whether IPYWIDGETS_ENABLED is true (inside the function body) and if not, fall back to _repr_.

This *would* be a sane approach but my understanding is that it's not feasible, since `_ipython_repr_` is supposed to work via side effects, and `__repr__` works by returning a value.

Here's the exact quote from [`_ipython_repr_`'s documentation](https://ipython.readthedocs.io/en/stable/config/integrating.html#MyObject._ipython_display_)
> Displays the object as a side effect; the return value is ignored. If this is defined, all other display methods are ignored.

Reply:
Good point -- to be honest, I got this idea from another place pointing out that any errors at all defaulted back to calling `_repr_`.  So in principle ... if we just threw an error it would likely swallow that and call `_repr_`...

Reply:
Minimal testing seems to indicate this doesn't work: trying to import `ipywidgets` if it's not installed *will* raise an exception (which is the original motivation for making it a hard dep). Or is there a special exception type we should raise instead ? Can you remember where this is coming from ?

Reply:
What about pointing `_ipython_display_` to a callable dynamically during `__init__` or (`__new__`?)? 

i.e., rename the current `_ipython_display_` to something else :

```python
    def _ipython_display_when_needed_(self): # or some other name
        import ipywidgets
        from IPython.display import display 
        ... etc ... 
```

Then in `__init__` or `__new__`:

```python
    def __init__(self):         
        if IPYWIDGETS_ENABLED:
            setattr(self, '_ipython_display_', '_ipython_display_when_needed_')
``` 

this would avoid the conditional at the class body level and would avoid the import error when ipywidget is not installed.

Reply:
This works. In my opinion it's less clear but maybe that's a possible compromise. What do you think @matthewturk ?

Reply:
I want to take a step back and say, I don't think that my disinclination is necessarily a dealbreaker.  Already we monkey-patch some stuff, which is just the same concept but at a different level.  So, whatever folks are OK with.

Reply:
@chrishavlin can have the final say as far as I'm concerned :)

Reply:
hah, ok :) then IMO I think sticking with what you already have is the way to go -- I do agree with Matt that having the conditional logic at the class level is not ideal. it does make it a bit harder to read what's going on there... but I do think it will make it easier to maintain going forward compared to my suggestion of dynamically linking the function, which kinda hides the logic away.

Reply:
though i do think adding a comment as to why it's at the class level would be helpful in the future.
 
```suggestion
    if IPYWIDGETS_ENABLED:
        # for discussion of this class-level conditional: https://github.com/yt-project/yt/pull/4941
```